File: step01activity/src/main/java/com/example/step01activity/MainActivity.java
package com.example.step01activity;

import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

//app 을 실행했을때 처음 사용자를 대면하는 MainActivity
public class MainActivity extends AppCompatActivity {

    //액티비티가 활성화 될때 onCreate() 메소드가 호출된다.
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //res/layout/activity_main.xml 문서를 전개해서 화면 구성하기
        setContentView(R.layout.activity_main);
        //액티비티가 활성화 되는 시점에 원하는 작업이 있으면 여기에 코딩한다.
    }

    /*
        activity_main.xml 에 정의된 버튼을 눌렀을때 호출되는 메소드 정의하기
        메소드의 인자로 View type 을 전달받도록 만들어야 한다.
     */
    public void sendClicked(View v){
        //콘솔창에 문자열 출력하기
        Log.i("sendClicked()", "전송 버튼을 눌렀네?");
        //잠깐 떳다가 사라지는 토스트 메세지 띄우기
        Toast.makeText(this, "전송 버튼을 눌렀네?", Toast.LENGTH_LONG).show();
    }
    public void deleteClicked(View v){
        //알림 다이얼로그 띄우기
        new AlertDialog.Builder(this)
                .setTitle("알림")
                .setMessage("삭제 합니다")
                .setPositiveButton("확인", null)
                .create()
                .show();

        /*위의 코드를 여러줄로 작성하면 아래와 같다
        AlertDialog.Builder builder=new AlertDialog.Builder(this);
        builder.setTitle("알림");
        builder.setMessage("삭제 합니다.");
        builder.setPositiveButton("확인", null);
        AlertDialog dialog=builder.create();
        dialog.show();
        */
    }
}

File: step23notification/src/main/java/com/example/step23notification/MainActivity.kt
package com.example.step23notification

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat

class MainActivity : AppCompatActivity(), View.OnClickListener{
    //EditText 의 참조값을 저장할 필드
    lateinit var inputMsg:EditText

    //알림 채널 (중복되는 이름이 없도록 페키지명을 같이 넣어준다)
    val CHANNEL_ID="com.example.step23notification.MY_CHANNEL"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //EditText 의 참조값
        inputMsg=findViewById(R.id.inputMsg)

        //버튼의 참조값 얻어와서 리스너 등록
        val notiBtn: Button =findViewById(R.id.notiBtn)
        notiBtn.setOnClickListener(this)
    }

    override fun onClick(v: View?) {
        //입력한 문자열을 읽어와서
        var msg=inputMsg.text.toString().trim()
        //알림 띄우기
        makeAutoCancelNoti(msg)
    }
    //인자로 전달되는 문자열을 알림에 띄우는 함수
    fun makeAutoCancelNoti(msg:String){
        //이 앱의 알림 체널 만들기
        createNotificationChannel()
        //알림을 클릭했을때 실행할 Activity 정보를 가지고 있는 Intent 객체
        val intent = Intent(this, DetailActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            putExtra("msg", msg)
        }
        //아이디 얻어내기
        val currentId:Int = (System.currentTimeMillis()/1000).toInt()

        //Intent 객체를 인텐트 전달자 객체에 담는다.
        val pendingIntent: PendingIntent =
                PendingIntent.getActivity(this, currentId, intent, 0)

        //NotificationCompat.Builer 객체 생성
        val builder = NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(android.R.drawable.ic_btn_speak_now) //알림 아이콘
                .setContentTitle("오빠 나야~") //알림 제목
                .setContentText(msg) //알림 내용
                .setPriority(NotificationCompat.PRIORITY_DEFAULT) //우선순위
                .setContentIntent(pendingIntent) //인텐트 전달자 객체
                .setAutoCancel(true) //자동으로 알림이 없어지도록



        with(NotificationManagerCompat.from(this)) {
            // notificationId is a unique int for each notification that you must define
            notify(currentId, builder.build())
        }
    }
    /*
        앱의 사용자가 알림을 직접 관리 할수 있도록 알림 채널을 만들어야한다.
     */
    private fun createNotificationChannel() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = "kim"
            val descriptionText = "test!"
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(CHANNEL_ID, name, importance).apply {
                description = descriptionText
            }
            // Register the channel with the system
            val notificationManager: NotificationManager =
                    getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(channel)
        }
    }

    //인자로 전달되는 문자열을 알림에 띄우는 함수
    fun makeManualCancelNoti(msg:String){
        //이 앱의 알림 체널 만들기
        createNotificationChannel()

        val currentId:Int = (System.currentTimeMillis()/1000).toInt()

        //알림을 클릭했을때 실행할 Activity 정보를 가지고 있는 Intent 객체
        val intent = Intent(this, DetailActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            putExtra("msg", msg)
            putExtra("id", currentId)
        }
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        intent.putExtra("msg", msg)

        //Intent 객체를 인텐트 전달자 객체에 담는다.
        val pendingIntent: PendingIntent =
                PendingIntent.getActivity(this, currentId, intent, 0)

        //NotificationCompat.Builer 객체 생성
        val builder = NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(android.R.drawable.ic_btn_speak_now)
                .setContentTitle("오빠 나야~")
                .setContentText(msg)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setContentIntent(pendingIntent)
                .setAutoCancel(false) //자동 cancel 되지 않도록

        with(NotificationManagerCompat.from(this)) {
            //수동으로 cancel 하려면 아래에 전달된 아이디 값을 알아야 한다.
            notify(currentId, builder.build())
        }
    }

}